src/spi.rs

```diff
@@ -0,0 +1,495 @@
+//! Serial Peripheral Interface (SPI) module.
+
+use ::nb;
+use ::hal::spi::{FullDuplex, Mode, Phase, Polarity};
+use ::stm32l4x6::{SPI1, SPI2, SPI3};
+
+use ::time::Hertz;
+use ::rcc::{APB1, APB2, Clocks};
+
+use ::ptr;
+
+///Describes GPIO Pins to be used by SPI.
+pub mod gpio {
+    ///SPI1 Pins
+    ///
+    ///Uses AF5
+    pub mod spi1 {
+        pub mod nss {
+            pub use ::gpio::{PA4, PA15};
+            #[cfg(feature = "STM32L476VG")]
+            pub use ::gpio::stm32l476vg::gpio::{PE12};
+        }
+        pub mod sck {
+            pub use ::gpio::{PA5, PB3};
+            #[cfg(feature = "STM32L476VG")]
+            pub use ::gpio::stm32l476vg::gpio::{PE13};
+        }
+        pub mod miso {
+            pub use ::gpio::{PA6, PB4};
+            #[cfg(feature = "STM32L476VG")]
+            pub use ::gpio::stm32l476vg::gpio::{PE14};
+        }
+        pub mod mosi {
+            pub use ::gpio::{PA7, PB5};
+            #[cfg(feature = "STM32L476VG")]
+            pub use ::gpio::stm32l476vg::gpio::{PE15};
+        }
+    }
+
+    ///SPI2 Pins
+    ///
+    ///Uses AF5
+    pub mod spi2 {
+        pub mod nss {
+            pub use ::gpio::{PB9, PB12};
+        }
+        pub mod sck {
+            pub use ::gpio::{PB10, PB13};
+        }
+        pub mod miso {
+            pub use ::gpio::{PB14, PC2};
+        }
+        pub mod mosi {
+            pub use ::gpio::{PB15, PC3};
+        }
+    }
+
+    ///SPI3 Pins
+    ///
+    ///Uses AF6.
+    ///Can overlap with others.
+    pub mod spi3 {
+        pub mod nss {
+            pub use ::gpio::{PA4, PA15};
+            #[cfg(feature = "STM32L476VG")]
+            pub use ::gpio::stm32l476vg::gpio::{PG12};
+        }
+
+        pub mod sck {
+            pub use ::gpio::{PB3, PC10};
+            #[cfg(feature = "STM32L476VG")]
+            pub use ::gpio::stm32l476vg::gpio::{PG9};
+        }
+
+        pub mod miso {
+            pub use ::gpio::{PB4, PC11};
+            #[cfg(feature = "STM32L476VG")]
+            pub use ::gpio::stm32l476vg::gpio::{PG10};
+        }
+
+        pub mod mosi {
+            pub use ::gpio::{PB15, PC12};
+            #[cfg(feature = "STM32L476VG")]
+            pub use ::gpio::stm32l476vg::gpio::{PG11};
+        }
+    }
+}
+
+use self::gpio::spi1::sck::*;
+use self::gpio::spi1::miso::*;
+use self::gpio::spi1::mosi::*;
+use self::gpio::spi2::sck::*;
+use self::gpio::spi2::miso::*;
+use self::gpio::spi2::mosi::*;
+use self::gpio::spi3::sck::*;
+use self::gpio::spi3::miso::*;
+use self::gpio::spi3::mosi::*;
+
+pub use ::gpio::{AF5, AF6};
+
+///Describes SCK Pin
+pub trait SCK {
+    ///SPI index
+    const SPI_IDX: u8;
+}
+
+///Describes MISO Pin
+pub trait MISO {
+    ///SPI index
+    const SPI_IDX: u8;
+}
+
+///Describes MOSI Pin
+pub trait MOSI {
+    ///SPI index
+    const SPI_IDX: u8;
+}
+
+macro_rules! impl_pins_trait {
```

Comment:
The AF and pin mux specifics vary a lot from chip to chip - pull the chip-specific stuff into its own module. Because module scope doesn't matter for impls (impls have a "global" effect), you can let each chip stand on its own. You can see here how I [separate out the chip cfg](https://github.com/thenewwazoo/stm32l4x6_hal/blob/serial-and-dma/src/serial/mod.rs#L17) and [how each chip gets its own impls](https://github.com/thenewwazoo/stm32l4x6_hal/blob/serial-and-dma/src/serial/stm32l496ag.rs).